api-processor/src/main/java/com/synopsys/integration/alert/processor/api/filter/JobNotificationFilterUtils.java

```diff
@@ -48,21 +50,26 @@ public static boolean doesNotificationTypeMatch(FilteredDistributionJobResponseM
         return filteredDistributionJobResponseModel.getNotificationTypes().contains(notificationType);
     }
 
-    public static boolean doesProjectApplyToJob(FilteredDistributionJobResponseModel filteredDistributionJobResponseModel, String projectName) {
+    public static boolean doesProjectApplyToJob(FilteredDistributionJobResponseModel filteredDistributionJobResponseModel, String projectName, String projectVersionName) {
         if (!filteredDistributionJobResponseModel.isFilterByProject()) {
             return true;
         }
 
         String projectNamePattern = filteredDistributionJobResponseModel.getProjectNamePattern();
-        if (projectNamePattern != null && Pattern.matches(projectNamePattern, projectName)) {
-            return true;
+        boolean matchingProjectNamePattern = (StringUtils.isNotBlank(projectNamePattern)) ? Pattern.matches(projectNamePattern, projectName) : false;
+        boolean hasMatchingProjects = filteredDistributionJobResponseModel.getProjectDetails()
+            .stream()
+            .map(BlackDuckProjectDetailsModel::getName)
+            .distinct()
+            .anyMatch(projectName::equals);
+
+        String projectVersionNamePattern = filteredDistributionJobResponseModel.getProjectVersionNamePattern();
+        if (StringUtils.isNotBlank(projectVersionNamePattern)) {
+            boolean listIsEmptyOrMatches = filteredDistributionJobResponseModel.getProjectDetails().isEmpty() || hasMatchingProjects;
```

Comment:
Shouldn't `hasMatchingProjects ` be false if the list is empty?

Reply:
oh true, this could have slipped in from previous iterations

Reply:
On further inspection, this case is important and changing this will fail a check in JobNotificationFilterUtilsTest. This check is here in the case that no selected projects match the notification project or no projects were ever selected. We then use this result to ensure the notification version matches as well.

Reply:
" in the case that no selected projects match the notification project or no projects were ever selected"
I'm still confused, is that not covered by `hasMatchingProjects `.

Reply:
I've updated the boolean with a comment. Because of our use cases, we care if the selected projects are empty or if they have a valid project. (true or true) If either of those are true, we then check the version pattern. If the selected projects are not empty and they do not have a valid project (false or false) then we have failed as we have to match an existing project if there is a list. Also, we could be empty and since we're empty, no valid project found (true or false) meaning the version pattern could still be true and we check.